### ISSUE TITLE
iFrame Title Attribute

### ISSUE THREAD

**Issue by TwisterMc** [ampproject/amp-wp#3313]
Currently I have a filter setup on `the_content` to add a title to all iFrames for accessibility purposes on non-AMP pages. However, AMP-WP strips out that title. I assume because the iFrame titles are not on the actual content.

Is there a way the AMP plugin can pull the iFrame title through to the AMP version if it exists on the non-AMP version? Or a way I can extend `AMP_Iframe_Sanitizer` to add in a title?

**Comment by westonruter** [comment]
I'm not seeing the `title` being stripped out.

I tried creating a Custom HTML block with this markup:

```html
<iframe title="Example Website" width="300" height="300" src="https://example.com/"></iframe>
```

And on an AMP page it is converted into:

```html
<amp-iframe title="Example Website" width="300" height="300" src="https://example.com/" sandbox="allow-scripts allow-same-origin" layout="intrinsic" class="amp-wp-enforced-sizes">
	<span placeholder="" class="amp-wp-iframe-placeholder"></span>
	<noscript>
		<iframe title="Example Website" width="300" height="300" src="https://example.com/"></iframe>
	</noscript>
</amp-iframe>
```

Perhaps you are referring to the `iframe` that is created by the `amp-iframe` component (in the light shadow DOM)?

![image](https://user-images.githubusercontent.com/134745/65344777-70bfcb00-db8d-11e9-8a28-1930e1940cf0.png)

**Comment by TwisterMc** [comment]
Technically you are correct. I guess it's not that WP-AMP is stripping it out rather than a situation where `the_content` is running on non-AMP posts and inserting a title into iFrames, but on AMP posts, I'm not getting a title added.

I'm assuming that when AMP looks for embedded YouTube videos (or other iframes) it's somehow going around the code we have working for non-AMP pages.

That _could_ be an issue with our code, but since we're running it on `the_content` I would assume AMP pages respect it.

Is that any more clear? :)

**Comment by westonruter** [comment]
Ah, so is this specifically for YouTube video iframes? Please share specific examples of the content and the code you're using to add the `title` attribute.

**Comment by TwisterMc** [comment]
Sorry for the lack of clarity, I thought all iFrames were the same.

The code in the original post is just a YouTube URL `https://www.youtube.com/watch?v=7aHq6tV-IcU`. WordPress then does its thing and embeds that video.

On page load, we have a filter on `the_content` that finds YouTube embeds and adds a title `title="YouTube Video"` to the `<iframe` code. 

I need to do something similar for AMP pages so that I can insert a title on those iFrames.

Does that help at all?

**Comment by TwisterMc** [comment]
I also did a second test where I disabled all my filters on `the_content` related to YouTube videos and the default functionality does not pull the iFrame title all the way through. 

Am I expecting functionality that doesn't exist in AMP? 

<img width="892" alt="iFrame Title Screenshot" src="https://user-images.githubusercontent.com/46997/65526886-69632f00-deb7-11e9-8324-65c994848ca9.png">

**Comment by westonruter** [comment]
> I also did a second test where I disabled all my filters on `the_content` related to YouTube videos and the default functionality does not pull the iFrame title all the way through.

It appears this is not necessary. I found this being raised in https://github.com/ampproject/amphtml/issues/17308#issuecomment-426399152:

> Your PR forwards the `title` attribute from the `amp-iframe` element to the `iframe` element, but this is unnecessary. The HTML specification (1) states that if an element's ancestor has a `title` attribute, that title applies to the child as well. This means that `title` is only needed on the `amp-iframe` element.
> 
> 1: https://html.spec.whatwg.org/multipage/dom.html#the-title-attribute

So duplicating the `title` attribute from the `amp-iframe` to the `iframe` appears to be unnecessary/redundant.

**Comment by westonruter** [comment]
Now, the reason why the `title` filter is not appearing in your YouTube videos is that the [`AMP_YouTube_Embed`](https://github.com/ampproject/amp-wp/blob/86370162531938d30d518ae9d552d6ba48a3b8ff/includes/embeds/class-amp-youtube-embed.php) class is hooking into the oEmbed functionality to cause an `amp-youtube` element to be output instead of an `iframe`:

https://github.com/ampproject/amp-wp/blob/86370162531938d30d518ae9d552d6ba48a3b8ff/includes/embeds/class-amp-youtube-embed.php#L135-L143

What needs to be done here is to do what was done for Soundcloud in #2722, and that is to preserve the `title` coming from the `iframe` provided by YouTube.

**Comment by westonruter** [comment]
In the short term then, what you can do is just have yout `the_content` filter look for `<amp-youtube>` elements in addition to `<iframe>` elements. Nevertheless, something I want to know is whether AMP `fallback` content is considered equivalent in any way to `title` attribute for purposes of a11y. Compare these two options:

```html
<amp-youtube data-videoid="tiQcA7cwEMo" layout="responsive" width="16" height="9" title="AMP for JavaScripters"></amp-youtube>
```

vs

```html
<amp-youtube data-videoid="tiQcA7cwEMo" layout="responsive" width="16" height="9">
    <a fallback href="https://youtube.com/watch?v=tiQcA7cwEMo">AMP for JavaScripters</a>
</amp-youtube>
```

Or also, what about both:

```html
<amp-youtube data-videoid="tiQcA7cwEMo" layout="responsive" width="16" height="9" title="AMP for JavaScripters">
    <a fallback href="https://youtube.com/watch?v=tiQcA7cwEMo">AMP for JavaScripters</a>
</amp-youtube>
```

In this last case, the dynamically-created `iframe` ancestor element will be the `amp-youtube` element with a `title`. Additionally, there would be a `fallback` sibling that gets `display: none`. So I think actually both are required. The `fallback` is useful for crawlers who aren't evaluating the JS for the `amp-youtube` component, or if the JS for the component fails to load.

**Comment by westonruter** [comment]
So I think I answered my own question 😄 This is the proposed result in #3358:

![image](https://user-images.githubusercontent.com/134745/65735102-c7943b80-e08a-11e9-9ea7-45fbeb1e6146.png)

**Comment by csossi** [comment]
Verified in QA
